ansible_base/lib/routers/association_resource_router.py

```diff
@@ -0,0 +1,187 @@
+        if non_related_instances:
+            raise serializers.ValidationError(
+                {
+                    'instances': _('Cannot disassociate these objects because they are not all related to this object: %(non_related_instances)s')
+                    % {'non_related_instances': ', '.join(str(i.pk) for i in non_related_instances)},
+                }
+            )
+
+        try:
+            manager.remove(*related_instances)
+        except AttributeError:
+            # If the field is FK and non-nullable, .remove() won't exist.
+            # https://docs.djangoproject.com/en/dev/ref/models/relations/#django.db.models.fields.related.RelatedManager.remove
+            # "For ForeignKey objects, this method only exists if null=True."
+            # Return something stating that the operation is not supported.
+            raise serializers.ValidationError(
+                {
+                    'instances': _('Cannot disassociate these objects because there must be a related object'),
+                }
+            )
+
+        return Response(status=status.HTTP_204_NO_CONTENT)
+
+    def get_serializer_class(self):
+        association_class = getattr(self, 'association_serializer', None)
+        if self.action in ('associate', 'disassociate'):
+            if association_class is None:
+                raise RuntimeError('You must set association_serializer on the viewset')
+            return association_class
+        return super().get_serializer_class()
+
+
+class ReverseViewMixin(QuerySetMixinBase):
+    def get_serializer_class(self):
+        related_class = getattr(self, 'related_serializer', None)
+        if self.action in ('associate', 'disassociate'):
+            if related_class is None:
+                raise RuntimeError('You must set related_serializer on the viewset')
+            return related_class
+        return super().get_serializer_class()
+
+
+class AssociationResourceRouter(routers.SimpleRouter):
+    def get_method_map(self, viewset, method_map):
+        is_associate_viewset = issubclass(viewset, AssociateMixin)
+        associate_actions = ['associate', 'disassociate', 'list']
+        bound_methods = {}
+        for method, action_str in method_map.items():
+            if hasattr(viewset, action_str):
+                if is_associate_viewset and action_str not in associate_actions:
+                    continue
+                bound_methods[method] = action_str
+        return bound_methods
+
+    def register(self, prefix, viewset, **kwargs):
+        basename = kwargs.get('basename', None)
+        related_views = kwargs.get('related_views', {})
+        reverse_views = kwargs.get('reverse_views', {})
+        if basename is None:
+            basename = self.get_default_basename(viewset)
+
+        for reverse_name, (reverse_view, fk) in reverse_views.items():
+
+            def related_serializer_factory(reverse_view=reverse_view):
```

Comment:
code style nit: better not to define the factory in-line. Yes, you absolutely need to define `RelatedSerializer` in-line. What I'm saying is that `def related_serializer_factory` does not need to be created inside the method here. It is much better to make that an instance method, class method, or totally stand-alone static method.

I realize you need `reverse_view`, but you made it an arg, so pass it as an arg. Making it the kwarg default is eyebrow-raising.

Reply:
> Making it the kwarg default is eyebrow-raising.

You probably know this already, but for others wondering why it does that: It's actually really important to do that here, with the way this is currently defined due to Python's late binding rules. The code is _incorrect_ if the arg is removed, it will end up always referencing the last thing `reverse_view` was set to (in the last iteration of the loop) in every serializer it creates:

```python
In [5]: funcs = []

In [6]: for i in range(5):
   ...:     def foo():
   ...:         return i
   ...:     funcs.append(foo)
   ...: 

In [7]: for func in funcs:
   ...:     print(func())
   ...: 
4
4
4
4
4
```

Reply:
Function moved out.

Reply:
I merged the related and the reverse views and also got rid of the kwargs (I only added it to make a test easier but that went away). Take a look and let me know if you are good with the direction I'm going. There is a bug that I need to track down later this afternoon.